src/adapter/src/coord/sequencer/inner.rs

```diff
@@ -2221,114 +2202,95 @@ impl Coordinator {
+        // necessary to support PG's `BEGIN` semantics, whose behavior can
+        // depend on whether or not reads have occurred in the txn.
+        if matches!(session.transaction(), &TransactionStatus::InTransaction(_))
+            || when == &QueryWhen::Immediately
+        {
+            session.add_transaction_ops(TransactionOps::Peeks(
+                determination.timestamp_context.clone(),
+            ))?;
+        }
+
+        // If we are in a single statement transaction, there is no need to
+        // acquire read holds to prevent compaction as they will be released
+        // immediately following the completion of the transaction.
+        //
+        // If we're in a multi-statement transaction, acquire read holds on any sources
+        // in the current time-domain if they have not already been acquired.
+        if session.transaction().is_in_multi_statement_transaction() {
+            self.check_txn_timedomain_conflicts(
+                session,
+                &determination.timestamp_context,
+                id_bundle,
+            )?;
+
+            let timedomain_id_bundle =
+                self.timedomain_for(source_ids, &timeline_context, session.conn_id(), cluster_id)?;
+
+            // If we've already acquired read holds for the txn, we can skip doing so again
+            if !self.txn_reads.contains_key(&session.conn_id()) {
+                if let Some(timestamp) = determination.timestamp_context.timestamp() {
+                    let read_holds =
+                        self.acquire_read_holds(timestamp.clone(), &timedomain_id_bundle);
+                    self.txn_reads.insert(session.conn_id(), read_holds);
+                }
+            }
+        }
+
+        Ok(determination)
+    }
+
     fn plan_peek(
-        &self,
+        &mut self,
         mut dataflow: DataflowDescription<OptimizedMirRelationExpr>,
-        session: &Session,
+        session: &mut Session,
         when: &QueryWhen,
         cluster_id: ClusterId,
         view_id: GlobalId,
         index_id: GlobalId,
         timeline_context: TimelineContext,
         source_ids: BTreeSet<GlobalId>,
         id_bundle: CollectionIdBundle,
-        in_immediate_multi_stmt_txn: bool,
         real_time_recency_ts: Option<Timestamp>,
         key: Vec<MirScalarExpr>,
         typ: RelationType,
     ) -> Result<PlannedPeek, AdapterError> {
-        let mut read_holds = None;
         let conn_id = session.conn_id();
-        // For transactions that do not use AS OF, get the timestamp context of the
-        // in-progress transaction or create one. If this is an AS OF query, we
-        // don't care about any possible transaction timestamp context. If this is a
-        // single-statement transaction (TransactionStatus::Started), we don't
-        // need to worry about preventing compaction or choosing a valid
-        // timestamp context for future queries.
```

Comment:
@jkosh44 Something I updated from this refactor is that an `AS OF` query inside a txn will now place read holds on objects in its timedomain. It looks like previously this wasn't the case as acquiring read holds was only done if `when == QueryWhen::Immediately`

This seems right to me? But just wanted to point that out in case I'm missing something and this needs to be reverted to the original behavior

Reply:
Ah, that's fair. Reverted